### PR TITLE
Feat: Undo waive charge

### PR DESCRIPTION
**Pull request by BLasan**
## Description

See [Issue-592](https://issues.apache.org/jira/browse/FINERACT-592)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by thesmallstar** [comment]
Please review commented code

**Comment by BLasan** [comment]
> Please review commented code

Removed

**Comment by thesmallstar** [comment]
will review this today

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/exception/LoanChargeWaiveCannotBeReversedException.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    /*** enum of reasons of why Loan Charge waive cannot undo **/
+    public enum LoanChargeWaiveCannotUndoReason {
+
+        ALREADY_PAID, ALREADY_WAIVED, LOAN_INACTIVE, WAIVE_NOT_ALLOWED_FOR_CHARGE, NOT_WAIVED, ALREADY_REVERSED;
+
+        public String errorMessage() {
+            /**
+             * TODO: Recheck the error messages
+             */
```
Comment (thesmallstar) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/exception/LoanChargeWaiveCannotBeReversedException.java]: who needs to do this?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java [hunk 2]
```diff
@@ -255,11 +256,34 @@ public String adjustLoanTransaction(@PathParam("loanId") @Parameter(description
...
-        final CommandProcessingResult result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);
+        final CommandWrapper commandRequest = builder.adjustTransaction(loanId, transactionId).build();
+        result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);
 
         return this.toApiJsonSerializer.serialize(result);
     }
 
+    /**
+     * TODO: Add Swagger annotations.
+     *
```
Comment (thesmallstar) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java]: Is this done?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java [hunk 3]
```diff
@@ -255,11 +256,34 @@ public String adjustLoanTransaction(@PathParam("loanId") @Parameter(description
             @Parameter(hidden = true) final String apiRequestBodyAsJson) {
 
         final CommandWrapperBuilder builder = new CommandWrapperBuilder().withJson(apiRequestBodyAsJson);
-        final CommandWrapper commandRequest = builder.adjustTransaction(loanId, transactionId).build();
+        CommandProcessingResult result = null;
 
-        final CommandProcessingResult result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);
+        final CommandWrapper commandRequest = builder.adjustTransaction(loanId, transactionId).build();
+        result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);
```
Comment (thesmallstar) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java]: can you please explain this part?

**Comment by thesmallstar** [comment]
@awasum @ptuomola @vorburger Will request your review on this one once...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java (hunk 2 above)
Comment (BLasan) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java]: This is done. Will remove this

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/exception/LoanChargeWaiveCannotBeReversedException.java (hunk 1 above)
Comment (BLasan) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/exception/LoanChargeWaiveCannotBeReversedException.java]: This needs to be removed. Will do it

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java (hunk 3 above)
Comment (BLasan) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java]: This is adjusting the transaction. Earlier I tried to do undoing part using this API. Will refine this. I've already introduced one for that.

**Comment by awasum** [comment]
100+ Commits? And conflicts. Travis not green.

Please squash these commits.

**Comment by BLasan** [comment]
> 100+ Commits? And conflicts. Travis not green.
> 
> Please squash these commits.

Done

**Comment by thesmallstar** [comment]
@awasum looks good to you? Should we merge this?

**Comment by thesmallstar** [comment]
@awasum please find required details here: https://gist.github.com/BLasan/bd21154ab9cc69e996a9e5aa0395cc03#4-undoing-a-waive-charge

**Comment by awasum** [comment]
@francisguchie  and @bharathc27 Please can you help functionally test this and make sure it works as expected so that it can be merged? For any questions, just ask here.

**Review by awasum — CHANGES_REQUESTED** [review]
Please add a unit test and integration test to this so we make sure it works now and also will work in future.

**Comment by BLasan** [comment]
> Please add a unit test and integration test to this so we make sure it works now and also will work in future.

Wrote the basic integration test on undoing a waive charge for installment fees. https://github.com/apache/fineract/pull/1836

**Review by thesmallstar — APPROVED** [review]
LGTM, 
1. The functionality/frontend has been tested by Bharat 
2. We have another PR linked for integration tests 

Will merge this in a few days if you don't have any issues: @awasum @xurror @vorburger @ptuomola

**Comment by thesmallstar** [comment]
Merging this, Thanks benura for working on this and Bharat for testing it. 
We still have some small issues with test cases but benura is targeting to get it done by today/tomorrow. 
Tests PR: #1836

**Pull request merged**